File: app/src/main/java/com/bitflaker/lucidsourcekit/database/dreamjournal/daos/JournalEntryTagDao.java
package com.bitflaker.lucidsourcekit.database.dreamjournal.daos;

import static androidx.room.OnConflictStrategy.IGNORE;

import androidx.room.Dao;
import androidx.room.Delete;
import androidx.room.Insert;
import androidx.room.Query;

import com.bitflaker.lucidsourcekit.database.dreamjournal.entities.JournalEntryTag;

import java.util.List;

import io.reactivex.rxjava3.core.Completable;
import io.reactivex.rxjava3.core.Single;

@Dao
public interface JournalEntryTagDao {
    @Query("SELECT * FROM JournalEntryTag")
    Single<List<JournalEntryTag>> getAll();

    @Query("SELECT description FROM JournalEntryTag")
    Single<List<String>> getAllTagTexts();

    @Query("SELECT COUNT(*) FROM JournalEntryTag")
    Single<Integer> getTotalTagCount();

    @Query("SELECT * FROM JournalEntryTag WHERE description IN (:descriptions)")
    Single<List<JournalEntryTag>> getByDescription(List<String> descriptions);

    @Insert
    Single<List<Long>> insertAll(JournalEntryTag... tags);

    @Insert(onConflict = IGNORE)
    Single<List<Long>> insertAll(List<JournalEntryTag> journalEntryTags);

    @Delete
    Completable delete(JournalEntryTag tag);
}

File: app/src/main/java/com/bitflaker/lucidsourcekit/database/dreamjournal/daos/JournalEntryTagDao.kt
package com.bitflaker.lucidsourcekit.database.dreamjournal.daos

import androidx.room.Dao
import androidx.room.Delete
import androidx.room.Insert
import androidx.room.OnConflictStrategy
import androidx.room.Query
import com.bitflaker.lucidsourcekit.database.dreamjournal.entities.JournalEntryTag

@Dao
interface JournalEntryTagDao {
    @Query("SELECT * FROM JournalEntryTag")
    suspend fun getAll(): List<JournalEntryTag>

    @Query("SELECT description FROM JournalEntryTag")
    suspend fun getAllTagTexts(): List<String>

    @Query("SELECT COUNT(*) FROM JournalEntryTag")
    suspend fun getTotalTagCount(): Int

    @Query("SELECT * FROM JournalEntryTag WHERE description IN (:descriptions)")
    suspend fun getByDescription(descriptions: List<String>): List<JournalEntryTag>

    @Insert
    suspend fun insertAll(vararg tags: JournalEntryTag): List<Long>

    @Insert(onConflict = OnConflictStrategy.Companion.IGNORE)
    suspend fun insertAll(journalEntryTags: MutableList<JournalEntryTag>): List<Long>

    @Delete
    suspend fun delete(tag: JournalEntryTag)
}
